models/migrations/v1_21/v263.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package v1_21 //nolint
+
+import (
+	"context"
+	"fmt"
+
+	"code.gitea.io/gitea/models/db"
+	"code.gitea.io/gitea/modules/timeutil"
+
+	"xorm.io/xorm"
+)
+
+func AddBranchTable(x *xorm.Engine) error {
+	type Branch struct {
+		ID            int64
+		RepoID        int64  `xorm:"index UNIQUE(s)"`
+		Name          string `xorm:"UNIQUE(s) NOT NULL"`
+		CommitSHA     string
+		CommitMessage string `xorm:"TEXT"`
+		PusherID      int64
+		IsDeleted     bool
+		DeletedByID   int64
+		DeletedUnix   timeutil.TimeStamp
+		CommitTime    timeutil.TimeStamp // The commit
+		CreatedUnix   timeutil.TimeStamp `xorm:"created"`
+		UpdatedUnix   timeutil.TimeStamp `xorm:"updated"`
+	}
+
+	if err := x.Sync(new(Branch)); err != nil {
+		return err
+	}
+
+	if exist, err := x.IsTableExist("deleted_branches"); err != nil {
+		return err
+	} else if !exist {
+		return nil
+	}
+
+	type DeletedBranch struct {
+		ID          int64
+		RepoID      int64  `xorm:"index UNIQUE(s)"`
+		Name        string `xorm:"UNIQUE(s) NOT NULL"`
+		Commit      string
+		DeletedByID int64
+		DeletedUnix timeutil.TimeStamp
+	}
+
+	var adminUserID int64
+	has, err := x.Table("user").
+		Select("id").
+		Where("is_admin=?", true).
+		Asc("id"). // Reliably get the admin with the lowest ID.
+		Get(&adminUserID)
+	if err != nil {
+		return err
+	} else if !has {
+		return fmt.Errorf("no admin user found")
+	}
+
+	branches := make([]Branch, 0, 100)
+	if err := db.Iterate(context.Background(), nil, func(ctx context.Context, deletedBranch *DeletedBranch) error {
+		branches = append(branches, Branch{
+			RepoID:    deletedBranch.RepoID,
+			Name:      deletedBranch.Name,
+			CommitSHA: deletedBranch.Commit,
+			PusherID:  adminUserID,
+			// CommitMessage: string, FIXME: how to get this?
+			// CommitTime:		timeutil.TimeStamp, FIXME: how to get this?
```

Comment:
As long as git gc hasn't been done, I think it's still possible to get the commit by `deletedBranch.Commit`. But I wonder if it's necessary for migrating deleted branchs, maybe just leave it black and remove the `FIXME`?